Sources/Common/DataModel/DataSetAttributes/FieldData.js

```diff
@@ -90,36 +100,112 @@ function vtkFieldData(publicAPI, model) {
         let destArr = publicAPI.getArrayByName(arr.getName());
         if (!destArr) {
           if (fromId < 0 || fromId > arr.getNumberOfTuples()) {
+            // TBD: should this be a deep or a shallow copy?
             publicAPI.addArray(arr);
+            other.getAttributes(arr).forEach((attrType) => {
+              publicAPI.setAttribute(arr, attrType);
+            });
           } else {
             const ncomps = arr.getNumberOfComponents();
             let newSize = arr.getNumberOfValues();
             const tId = toId > -1 ? toId : fromId;
-            if (newSize < tId * ncomps) {
+            if (newSize <= tId * ncomps) {
               newSize = (tId + 1) * ncomps;
             }
             destArr = vtkDataArray.newInstance({
               name: arr.getName(),
               dataType: arr.getDataType(),
-              numberOfComponents: arr.getNumberOfComponents(),
-              size: newSize,
+              numberOfComponents: ncomps,
+              values: macro.newTypedArray(arr.getDataType(), newSize),
+              size: 0,
             });
-            destArr.setTuple(tId, arr.getTuple(fromId));
+            destArr.insertTuple(tId, arr.getTuple(fromId));
             publicAPI.addArray(destArr);
+            other.getAttributes(arr).forEach((attrType) => {
+              publicAPI.setAttribute(destArr, attrType);
+            });
           }
         } else if (
           arr.getNumberOfComponents() === destArr.getNumberOfComponents()
         ) {
           if (fromId > -1 && fromId < arr.getNumberOfTuples()) {
             const tId = toId > -1 ? toId : fromId;
-            destArr.setTuple(tId, arr.getTuple(fromId));
+            destArr.insertTuple(tId, arr.getTuple(fromId));
+          } else {
+            // if fromId and not provided, just copy all (or as much possible)
+            // of arr to destArr.
+            for (let i = 0; i < arr.getNumberOfTuples(); ++i) {
+              destArr.insertTuple(i, arr.getTuple(i));
```

Comment:
you may want to create a `getTuples()` convenient vtkDataArray method that returns a subarray